packages/utils/src/get-snark-artifacts/get-snark-artifacts.node.ts

```diff
@@ -0,0 +1,89 @@
+import { createWriteStream, existsSync } from "node:fs"
+import { mkdir } from "node:fs/promises"
+import { dirname } from "node:path"
+import os from "node:os"
+import { SnarkArtifacts, ProofType } from "../types"
+import { ARTIFACT_BASE_URLS, ARTIFACTS_TYPES } from "./config"
+
+const getArtifactBaseUrl = (proofType: ProofType, numberOfInputs?: number): string => {
+    if (proofType === ProofType.POSEIDON) {
+        if (numberOfInputs === undefined) {
+            throw new Error("numberOfInputs is required for Poseidon proof")
+        }
+        if (numberOfInputs < 1) {
+            throw new Error("numberOfInputs must be greater than 0")
+        }
+        return `${ARTIFACT_BASE_URLS[proofType]}/${numberOfInputs}`
+    }
+    return ARTIFACT_BASE_URLS[proofType]
+}
+
+async function download(url: string, outputPath: string) {
+    const response = await fetch(url)
+    if (!response.ok) {
+        throw new Error(`Failed to fetch ${url}: ${response.statusText}`)
+    }
+    if (!response.body) {
+        throw new Error("Failed to get response body")
+    }
+
+    const dir = dirname(outputPath)
+    await mkdir(dir, { recursive: true })
+
+    const fileStream = createWriteStream(outputPath)
+    const reader = response.body.getReader()
+
+    try {
+        const pump = async () => {
+            const { done, value } = await reader.read()
+            if (done) {
+                fileStream.end()
+                return
+            }
+
+            fileStream.write(Buffer.from(value))
+            await pump()
+        }
+
+        await pump()
+    } catch (error) {
+        fileStream.close()
+        throw error
+    }
```

Comment:
even if node > 18 has built in fetch, there are still some mismatch between the stream api of node and [whatwg](https://streams.spec.whatwg.org/): node fetch api returns a [ReadableStream](https://streams.spec.whatwg.org/#rs-class) which isn't directly compatible with node stream system. This is why `as any` type casting was necessary.
We can avoid type casting by using the stream reader from the ReadableStream which is part of the WHATWG api.  
It also makes unit testing easier.  
But it is lot more verbose....

Reply:
Sounds good to me! How much of a decrease in performance?

Reply:
Actually, and unexpectedly, an increase of performance (probably due to the fact that I am fetching with parallel with `Promise.all` too?)

benchmark on poseidon. Run it 2 times. Gave similar results both times
```
                   pipe pump
poseidon 1 inputs: 2389 44
poseidon 2 inputs: 690 1
poseidon 3 inputs: 723 1
poseidon 4 inputs: 672 1
poseidon 5 inputs: 707 3
poseidon 6 inputs: 915 1
poseidon 7 inputs: 1653 1
poseidon 8 inputs: 722 1
poseidon 9 inputs: 878 1
poseidon 10 inputs: 772 1
poseidon 11 inputs: 790 1
poseidon 12 inputs: 806 1
poseidon 13 inputs: 819 1
poseidon 14 inputs: 851 1
poseidon 15 inputs: 776 1
poseidon 16 inputs: 871 1
```

Reply:
Ohh great news 👍🏽